sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ToNumberParser.scala

```diff
@@ -0,0 +1,517 @@
+    // Make sure that any thousands separators in the format string have digits before and after.
+    else if (digitGroupsBeforeDecimalPoint.exists {
+      case DigitGroups(tokens) =>
+        tokens.zipWithIndex.exists({
+          case (_: ThousandsSeparator, j: Int) if j == 0 || j == tokens.length - 1 =>
+            true
+          case (_: ThousandsSeparator, j: Int) if tokens(j - 1).isInstanceOf[ThousandsSeparator] =>
+            true
+          case (_: ThousandsSeparator, j: Int) if tokens(j + 1).isInstanceOf[ThousandsSeparator] =>
+            true
+          case _ =>
+            false
+        })
+    }) {
+      "Thousands separators (,) must have digits in between them " +
+        s"in the number format: '$originNumberFormat'"
+    }
+    // Thousands separators are not allowed after the decimal point, if any.
+    else if (digitGroupsAfterDecimalPoint.exists {
+      case DigitGroups(tokens) => tokens.exists(_.isInstanceOf[ThousandsSeparator])
+    }) {
+      "Thousands separators (,) may not appear after the decimal point " +
+        s"in the number format: '$originNumberFormat'"
+    }
+    // Validation of the format string finished successfully.
+    else {
+      ""
+    }
+  }
+
+  /**
+   * Convert string to numeric based on the given number format.
+   *
+   * Iterates through the [[inputTokens]] obtained from processing the format string, while also
+   * keeping a parallel index into the input string. Throws an exception if the latter does not
+   * contain expected characters at any point.
+   *
+   * @param input the string that needs to converted
+   * @return the result Decimal value obtained from string parsing
+   */
+  def parse(input: UTF8String): Decimal = {
+    val inputStr = input.toString.trim
+    val beforeDecimalPoint = new StringBuilder()
+    val afterDecimalPoint = new StringBuilder()
+    var reachedDecimalPoint = false
+    // Number of times that the input specified a negative result, such as with a minus sign.
+    var negateResult = 0
+    // This is an index into the characters of the provided input string.
+    var inputIndex = 0
+    // This is an index into the tokens of the provided format string.
+    var formatIndex = 0
+
+    // Iterate through the tokens representing the provided format string, in order.
+    while (formatIndex < inputTokens.size) {
+      val token: InputToken = inputTokens(formatIndex)
+      token match {
+        case d@DigitGroups(_) =>
+          val expectedTokens: Seq[InputToken] = d.tokens.filter(_.isInstanceOf[Digits])
+          // Consume characters from the current input index forwards in the input string as long as
+          // they are digits (0-9) or the thousands separator (,). Then split these characters into
+          // groups by the thousands separator (,). For example, the input string
+          // "456,789" becomes the actualTokens array ("789", "456").
+          val actualTokens: Array[String] = inputStr.drop(inputIndex).takeWhile {
```

Comment:
let's avoid calling `.drop` in the loop iteration as it creates a new `Seq` instance.

Reply:
Good point. This runs on a tight loop for every input row. I updated the implementation to completely avoid allocating or copying memory now. This should be faster.